Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -83,74 +87,166 @@ public async Task<BoolResult> ShutdownAsync(Context context)
+                            break;
+                    }
+                }
+
+                return CopyFileResult.Success;
             }
-            catch (RpcException r) when (r.StatusCode == StatusCode.Unavailable && r.Message.Contains("Connect Failed"))
+            catch (RpcException r)
             {
-                return new CopyFileResult(CopyFileResult.ResultCode.SourcePathError, r, $"Failed to connect to server {_host} at port {_grpcPort}");
+                if (r.StatusCode == StatusCode.Unavailable)
+                {
+                    return new CopyFileResult(CopyFileResult.ResultCode.SourcePathError, r);
+                }
+                else
+                {
+                    return new CopyFileResult(CopyFileResult.ResultCode.Unknown, r);
+                }
             }
 
-            if (bytesReceived == 0)
-            {
-                return new CopyFileResult(CopyFileResult.ResultCode.SourcePathError, $"Received {bytesReceived} bytes for {contentHash}. Source file does not exist");
-            }
-            else if (fileSize >= 0 && bytesReceived != fileSize)
+        }
+
+        private async Task<(long, long)> StreamContentAsync(Stream targetStream, IAsyncStreamReader<CopyFileResponse> replyStream, CancellationToken ct = default(CancellationToken))
+        {
+            long chunks = 0L;
+            long bytes = 0L;
+            while (await replyStream.MoveNext(ct).ConfigureAwait(false))
             {
-                return new CopyFileResult(CopyFileResult.ResultCode.InvalidHash, $"Received {bytesReceived} bytes for {contentHash}, expected {fileSize}");
+                chunks++;
+                CopyFileResponse reply = replyStream.Current;
+                bytes += reply.Content.Length;
+                reply.Content.WriteTo(targetStream);
             }
-
-            return CopyFileResult.SuccessWithSize(bytesReceived);
+            return (chunks, bytes);
         }
 
-        private async Task<long> StreamContentAsync(Stream fileStream, IAsyncStreamReader<CopyFileResponse> replyStream)
+        private async Task<(long, long)> StreamContentWithCompressionAsync(Stream targetStream, IAsyncStreamReader<CopyFileResponse> replyStream, CancellationToken ct = default(CancellationToken))
         {
-            long bytesReceived = 0L;
-            while (await replyStream.MoveNext(CancellationToken.None))
+            Debug.Assert(targetStream != null);
+            Debug.Assert(replyStream != null);
+
+            long chunks = 0L;
+            long bytes = 0L;
+            using (BufferedReadStream grpcStream = new BufferedReadStream(async () =>
             {
-                CopyFileResponse oneOfManyReply = replyStream.Current;
-                bytesReceived += oneOfManyReply.Content.Length;
-                oneOfManyReply.Content.WriteTo(fileStream);
+                if (await replyStream.MoveNext(ct).ConfigureAwait(false))

```

Comment:
Would a cancellation token check make sense here?

Reply:
GRPC's MoveNext(ct) cancels if ct.IsCancellationRequested AND transmits that cancellation to the service, so I actually want the ct to get down into that method.